Stats doc (#185)
opendatacube/odc-tools: libs/stats/README.md

```
odc-stats save-tasks --frequency annual --temporal-range 2015--P6Y --grid au-10  ga_ls8c_ard_3
```

The above command will generate the following files i.e. a csv file containing a list of tasks for years between 2015 to 2020, a cache file that will be used as an input to the stats cli, and several geojson files (one for each year).  

```
ga_ls8c_ard_3_2015--P6Y-2015--P1Y.geojson  
ga_ls8c_ard_3_2015--P6Y-2016--P1Y.geojson  
ga_ls8c_ard_3_2015--P6Y-2017--P1Y.geojson  
ga_ls8c_ard_3_2015--P6Y-2018--P1Y.geojson
ga_ls8c_ard_3_2015--P6Y-2019--P1Y.geojson
ga_ls8c_ard_3_2015--P6Y-2020--P1Y.geojson  
ga_ls8c_ard_3_2015--P6Y.db
ga_ls8c_ard_3_2015--P6Y.csv
```

The csv file contains the list of tasks for all the years and consists of x, y coordinates of each tile to be processed as well as the counts of datasets and satellite observations.

geojson files are useful when selecting test regions as well as for debugging specific tasks - see the example below from Landsat-8, 2015.  This example shows a tile from Australia coastal region near Sydney.

<img src="odc/stats/stats/auxiliary/screenshot-L8-2015.png" alt="drawing" width="1000"/>

### 2- Run stats
Sample command:
```

> odc-stats run s3-path/ga_ls8c_nbart_gm_cyear_3.db  2015--P1Y/41/13 --threads=16 --memory-limit=60Gi --resolution=30 --config cfg.txt --location file://localpath/

Reviewer: file urls must have 3 forward slashes, `file:///`

Author: Done.

> odc-stats run s3-path/ga_ls8c_nbart_gm_cyear_3.db  2015--P1Y/41/13 --threads=16 --memory-limit=60Gi --resolution=30 --config cfg.txt --location file://localpath/

Reviewer: Let's get rid of `s3-path/` prefix, and make sure that db name is the same as was produced by the save-tasks step.

Author: Addressed accordingly.

> odc-stats run s3-path/ga_ls8c_nbart_gm_cyear_3.db  2015--P1Y/41/13 --threads=16 --memory-limit=60Gi --resolution=30 --config cfg.txt --location file://localpath/

Reviewer: config is in yaml format not text

Author: fixed.

Usage
-----
Stats offers a set of tools to generate clear pixel count and geometric median on Sentinel-2 and Landsat satellite images and it's currently being extended to support sibling products.

> ### Steps to run stats

Reviewer 2: Can we call it Statistician?

### 1- Save tasks

> From your sandbox (or a machine that has access to your database), run:

Reviewer 2: ```suggestion
From your Sandbox (or a machine that has access to an ODC database), run:
```

```
odc-stats save-tasks --frequency annual --grid au-30  ga_ls8c_ard_3
```

The above command will generate the following files i.e. a csv file containing a list of tasks for all the years in the database, a cache file that will be used as an input to the stats cli, and several geojson files (one for each year).  

```
ga_ls8c_ard_3_all-2013--P1Y.geojson
ga_ls8c_ard_3_all-2014--P1Y.geojson
ga_ls8c_ard_3_all-2015--P1Y.geojson
ga_ls8c_ard_3_all-2016--P1Y.geojson
ga_ls8c_ard_3_all-2017--P1Y.geojson
ga_ls8c_ard_3_all-2018--P1Y.geojson
ga_ls8c_ard_3_all-2019--P1Y.geojson
ga_ls8c_ard_3_all-2020--P1Y.geojson
ga_ls8c_ard_3_all-2021--P1Y.geojson
ga_ls8c_ard_3_all.csv
ga_ls8c_ard_3_all.db
```

The csv file contains the list of tasks for all the years and consists of x, y coordinates of each tile to be processed as well as the counts of datasets and satellite observations.

> geojson files are useful when selecting test regions as well as for debugging specific tasks - see the example below from Landsat-8, 2015.  This example shows a tile from Australia coastal region near Sydney.

Reviewer 2: ```suggestion
GeoJSON files are useful when selecting test regions as well as for debugging specific tasks - see the example below from Landsat-8, 2015.  This example shows a tile from an Australian coastal region near Sydney.
```

<img src="odc/stats/stats/auxiliary/screenshot-L8-2015.png" alt="drawing" width="1000"/>

### 2- Run stats
Sample command:
```
odc-stats run ga_ls8c_ard_3_all.db  2015--P1Y/41/13 --threads=16 --memory-limit=60Gi --resolution=30 --config cfg.yaml --location file:///localpath/
```  

> Where cfg.yaml contains the following configurations (sample from a geomedian run):

Reviewer 2: ```suggestion
Where cfg.yaml contains the following configuration (sample from a geomedian run):
```

```
plugin: gm-ls
plugin_config:
  resampling: bilinear
  bands: ["blue", "red", "green", "nir", "swir1", "swir2"]
  rgb_bands: ["red", "green", "blue"]
  mask_band: fmask
  cloud_classes: ["shadow", "cloud"]
  filters: [0, 0] 
  aux_names:
    smad: sdev
    emad: edev
    bcmad: bcdev
    count: count                   
product:
  name: ga_ls8c_nbart_gm_cyear_3
  short_name: ga_ls8c_nbart_gm_cyear_3 
  version: 3.0.0
  product_family: geomedian
max_processing_time: 3600
job_queue_max_lease: 300
renew_safety_margin: 60
future_poll_interval: 2
s3_acl: public-read
# Generic product attributes
cog_opts:
  zlevel: 9
  overrides:
    rgba:
      compress: JPEG
      jpeg_quality: 90
```

Note that the configurations vary between different products.  A sample plugin configuration for Australia on Landsat-8 can be found here: 
```

> https://bitbucket.org/geoscienceaustralia/datakube-apps/src/develop/workspaces/dea-dev/processing/06_stats.yaml

Reviewer 2: This is not public...

Can we include some example configuration in this repo?

```

Orchestration
-----

In order to run stats on a large volume of data, use stats orchestration tool which relies on SQS queues and Kubernetes jobs.  Before running a job, one needs to create the necessary infrastructures by running the step 1 described below.

> ### 1- Create queues and a user 

Reviewer 2: Hmm... a lot of this is DEA specific. Probably needs @whatnick or @NikitaGandhi to suggest the right way to document?

### Issues 

During the run, monitor the number of inflight messages.  This can be done from the cli or the dashboard.  Counts of inflight messages should be the same as (or close to) the number of pod/replicas.

Currently when the jobs complete, pods do not shutdown cleanly due to an error in the dask library.  In order to ensure assigned EC2 instances are released, you need to delete the job manually by running the following command:

``` 
kp delete -f path-to-job-template.yaml
```

Monitor the AWS Autoscaling Groups and ensure they are reduced to 0 once the job has completed.  There is a delay of about 10 minutes before ASGs are released. 

### 4- Redrive queue
For tasks that have failed and need to be retried, they can be redirected to the main stats queue using the following command:

```
redrive-queue stats-dead-letter-queue-name  dead-letter-queue-name 
```

> ### Monitoring

Reviewer 2: This is DEA specific too...

Usage
-----
Stats offers a set of tools to generate clear pixel count and geometric median on Sentinel-2 and Landsat satellite images and it's currently being extended to support sibling products.

### Steps to run stats

### 1- Save tasks

From your sandbox (or a machine that has access to your database), run:

```

> odc-stats save-tasks --frequency annual --temporal-range 2015--P6Y --grid au-10  ga_ls8c_ard_3

Reviewer: Let's get rid off `--temporal-range`, also we should use `au-30` for landsat.

```

The above command will generate the following files i.e. a csv file containing a list of tasks for years between 2015 to 2020, a cache file that will be used as an input to the stats cli, and several geojson files (one for each year).  

```

> ga_ls8c_ard_3_2015--P6Y-2015--P1Y.geojson  

Reviewer: this looks broken

Author: Removed the frequency and updated the geojson files.

Usage
-----
Stats offers a set of tools to generate clear pixel count and geometric median on Sentinel-2 and Landsat satellite images.
### Steps to run stats

### 1- Save tasks

From a machine that has access to your database, run:

```

> odc-stats save-tasks --year YYYY --grid  grid-name  --temporal-range range product-name

Reviewer: Let's use real examples, examples should be able to run in AU sandbox.

```

> This will generate a csv file containing the list of tasks to be processed, a cache file that will be used as an input to the stats cli, and geojson files.  

Reviewer: Also need to mention `.geojson` files that get generated. These can be used to find a labels for test region. Some screenshots would be nice to have too.

> ### 2- Publish tasks 

Reviewer: This section should be AFTER "run section". Let's first describe how to test configuration/code, and only then focus on large scale processing.
